### PR TITLE
Fix reset GUC issue that causes inconsistent GUC values

### PR DESCRIPTION
**Pull request by gfphoenix78**
When we change some GUC values and reset them later in
a session. The GUC values may be inconsistent among QD and/or QEs.
The key steps of current GUC settings are:
1. The QD dispatches its current `sync`ed GUC values(as startup
    options) to create QEs.
2. The QE sets the GUCs' `reset_val` to the options passsed
    from the QD during its process initialization.
2.5 When the QD dispatches `SET XXX=YYY` for an existing QE,
    the QE will set the GUC's current value to YYY, `reset_val`
    is unchanged. Because its guc-source is PGC_S_SESSION.
3. When the QE receives `RESET XXX`, it resets the GUC's value to
    its `reset_val`.

The problem happens in the step 1, which causes the `reset_val`s
between QD and QEs inconsistent if the QD changed the GUC before
creating the QE(s).

The fix is to send the GUC's `reset_val` to create QEs, and we also
have another options(diff_options) to update the changed GUCs. So,
both `reset_val` and current value are consistent among QD and QEs
if the GUC needs to be synchronized.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by gfphoenix78** [comment]
I'll add test cases soon, so please ignore them.

**Comment by kainwen** [comment]
Pair with @gfphoenix78 and take a look at this. It LGTM.

But I still want to verify if a simple idea can work: can we just destroy all gangs when reset GUC?

File: src/interfaces/libpq/libpq-int.h [hunk 1]
```diff
@@ -409,6 +409,7 @@ struct pg_conn
 #endif
     char       *gpconntype; /* type of connection */
     char       *gpqeid;        /* MPP: session id & startup info for qExec */
+    char       *diffoptions;
```
Comment (l-wang) [inline src/interfaces/libpq/libpq-int.h:412]: Could you add comment here?

File: src/interfaces/libpq/fe-connect.c [hunk 2]
```diff
@@ -396,6 +396,10 @@ static const internalPQconninfoOption PQconninfoOptions[] = {
 		"connection type", "D", 10,
 	offsetof(struct pg_conn, gpconntype)},
 
+	{"diff_options", NULL, NULL, NULL,
+		"updated synced GUCs", "D", 80,
```
Comment (l-wang) [inline src/interfaces/libpq/fe-connect.c:400]: Why size 80? Will it be enough if multiple GUCs were set to non-default values before creating the QEs? Do we need test case for that or is that irreverent?

File: src/interfaces/libpq/libpq-int.h (hunk 1 above)
Comment (gfphoenix78) [reply src/interfaces/libpq/libpq-int.h:412]: Sure.

File: src/interfaces/libpq/fe-connect.c (hunk 2 above)
Comment (gfphoenix78) [reply src/interfaces/libpq/fe-connect.c:400]: The field name of this value is **`dispsize`**,
```c
  int     dispsize;   /* Field size in characters for dialog  */
```

It looks meaningless. I find no usage in both Postgres and Greenplum. This value doesn't constraint the length of `diff_options`.
I've tested in my local environment, it doesn't matter whether its value is 80 or 4.

Comment (l-wang) [reply src/interfaces/libpq/fe-connect.c:400]: Interesting! Without much context on this subject, I did find [one commit in Postgres](https://github.com/postgres/postgres/commit/f4051e363c1757a5fa05825a361d9dd0e54508bc) that updated the `dispsize` of the `sslmode` field. But if you have tested different values for `diff_options` we are probably fine.

Comment (gfphoenix78) [reply src/interfaces/libpq/fe-connect.c:400]: FYI, the similar conninfo option is **`options`** which is also variable-length string, the `dispsize` of "options" is 40.
```c
  {"options", "PGOPTIONS", DefaultOption, NULL,
    "Backend-Options", "", 40,
  offsetof(struct pg_conn, pgoptions)},
```

Comment (ashwinstar) [reply src/interfaces/libpq/fe-connect.c:400]: The commit message you pointed @l-wang itself says "I wouldn't be surprised if nobody is actually using these
values ..." and even other similar commit from Tom fixing it for some other field mentions the same. For fixed length components its easy to know the value is wrong, for variable length don't care as other options have similar thing. Seems this is all dead values and no one really uses them, for sure I doubt for GPDB.

Comment (l-wang) [reply src/interfaces/libpq/fe-connect.c:400]: I see. Thank you @ashwinstar for clarifying and @gfphoenix78 for verifying!

**Review by l-wang — COMMENTED** [review]
> But I still want to verify if a simple idea can work: can we just destroy all gangs when reset GUC?

LGTM. +1 if this is also verified.

**Comment by gfphoenix78** [comment]
> Pair with @gfphoenix78 and take a look at this. It LGTM.
> 
> But I still want to verify if a simple idea can work: can we just destroy all gangs when reset GUC?

One exception is that RESET is allowed to run in a transaction block, so destroying all gangs can't always work for RESET GUC.

**Pull request merged**